pkg/network/encoding/http.go

```diff
@@ -30,132 +29,67 @@ var (
 			return new(model.HTTPStats)
 		},
 	}
-)
-
-type httpEncoder struct {
-	aggregations   map[types.ConnectionKey]*aggregationWrapper
-	staticTags     map[types.ConnectionKey]uint64
-	dynamicTagsSet map[types.ConnectionKey]map[string]struct{}
 
-	orphanEntries int
-}
-
-// aggregationWrapper is meant to handle collision scenarios where multiple
-// `ConnectionStats` objects may claim the same `HTTPAggregations` object because
-// they generate the same connection key
-// TODO: we should probably revisit/get rid of this if we ever replace socket
-// filters by kprobes, since in that case we would have access to PIDs, and
-// could incorporate that information in the `types.ConnectionKey` struct.
-type aggregationWrapper struct {
-	*model.HTTPAggregations
-
-	// we keep track of the source and destination ports of the first
-	// `ConnectionStats` to claim this `HTTPAggregations` object
-	sport, dport uint16
-}
-
-func (a *aggregationWrapper) ValueFor(c network.ConnectionStats) *model.HTTPAggregations {
-	if a == nil {
-		return nil
+	httpStatusCodeMap = sync.Pool{
+		New: func() any {
+			m := make(map[int32]*model.HTTPStats_Data)
+			return &m
+		},
 	}
+)
 
-	if a.sport == 0 && a.dport == 0 {
-		// This is the first time a ConnectionStats claim this aggregation. In
-		// this case we return the value and save the source and destination
-		// ports
-		a.sport = c.SPort
-		a.dport = c.DPort
-		return a.HTTPAggregations
-	}
+type httpEncoder struct {
+	byConnection *USMConnectionIndex[http.Key, *http.RequestStats]
 
-	if c.SPort == a.dport && c.DPort == a.sport {
-		// We have have a collision with another `ConnectionStats`, but this is a
-		// legit scenario where we're dealing with the opposite ends of the
-		// same connection, which means both server and client are in the same host.
-		// In this particular case it is correct to have both connections
-		// (client:server and server:client) referencing the same HTTP data.
-		return a.HTTPAggregations
-	}
+	// cached object
+	aggregations *model.HTTPAggregations
 
-	// Return nil otherwise. This is to prevent multiple `ConnectionStats` with
-	// exactly the same source and destination addresses but different PIDs to
-	// "bind" to the same HTTPAggregations object, which would result in a
-	// overcount problem. (Note that this is due to the fact that
-	// `types.ConnectionKey` doesn't have a PID field.) This happens mostly in the
-	// context of pre-fork web servers, where multiple worker processes share the
-	// same socket
-	return nil
+	// list of *pointers* to maps so they can be returned to the pool
+	toRelease []*map[int32]*model.HTTPStats_Data
```

Comment:
we're never initializing the array, thus we're using the builtin allocation mechanism. Do we want to change it?